### PR TITLE
EE-1046: Add or update client contracts for bids. (Rust side)

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/EE-1046

- Refactored out some test contracts to remove their bid/delegatin calls, and use client contracts instead for this (this will be used in another PR with AssemblyScript client contracts)
- Removed `bond/unbond` contracts
- Added `add_bid`, `withdraw_bid`, `delegate`, `undelegate` contracts
- In addition withdraw_bid/undelegate contracts are saving the unbonding purses returned from the contract API, to let users automatically keep track of funds that will be delivered in future era.

### CODE REVIEW

File: smart_contracts/contracts/client/undelegate/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// to withdraw (of type `U512`).
+#[no_mangle]
+pub extern "C" fn call() {
+    let delegator = runtime::get_named_arg(ARG_DELEGATOR);
+    let validator = runtime::get_named_arg(ARG_VALIDATOR);
+    let amount = runtime::get_named_arg(ARG_AMOUNT);
+    let undelegate_purse = undelegate(delegator, validator, amount);
+
+    // Save the purse under account's coontext to keep track of the funds.
+    runtime::put_key(UNDELEGATE_PURSE, undelegate_purse.into());
```
Comment (EdHastingsCasperLabs) [inline smart_contracts/contracts/client/undelegate/src/main.rs:38]: If an account has undelegated previously and not yet drained the funds from that earlier purse, wouldn't this effectively cause those funds to be lost forever?

File: smart_contracts/contracts/client/withdraw-bid/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// Accepts a public key to be removed, and an amount to withdraw (of type `U512`).
+// Saves the withdrawn funds in the account's context to keep track of the funds.
+#[no_mangle]
+pub extern "C" fn call() {
+    let public_key = runtime::get_named_arg(ARG_PUBLIC_KEY);
+    let amount = runtime::get_named_arg(ARG_AMOUNT);
+    let withdraw_bid_purse = withdraw_bid(public_key, amount);
+
+    // Save the purse under account's coontext to keep track of the funds.
+    runtime::put_key(WITHDRAW_BID_PURSE, withdraw_bid_purse.into());
```
Comment (EdHastingsCasperLabs) [inline smart_contracts/contracts/client/withdraw-bid/src/main.rs:35]: same question as undelegate...

File: smart_contracts/contracts/client/undelegate/src/main.rs (hunk 1 above)
Comment (mpapierski) [reply smart_contracts/contracts/client/undelegate/src/main.rs]: Done. I added 3rd argument `unbond_purse` to withdraw_bid/undelegate API and `Option<URef>` in client contract which would default to `main_purse`

**Pull request merged**